crates/go/src/wit_runtime.go

```diff
@@ -0,0 +1,83 @@
+package wit_runtime
+
+import (
+	"fmt"
+	"runtime"
+	"unsafe"
+)
+
+func Allocate(pinner *runtime.Pinner, size, align uintptr) unsafe.Pointer {
+	pointer := allocateRaw(size, align)
+	pinner.Pin(pointer)
+	return pointer
+}
+
+func ceiling(n, d uintptr) uintptr {
+	var hasRemainder uintptr
+	if n%d == 0 {
+		hasRemainder = 0
+	} else {
+		hasRemainder = 1
+	}
+	return n/d + hasRemainder
+}
+
+func allocateRaw(size, align uintptr) unsafe.Pointer {
+	if size == 0 {
+		return unsafe.Pointer(uintptr(0))
+	}
+
+	switch align {
+	case 1:
+		return unsafe.Pointer(unsafe.SliceData(make([]uint8, size)))
+	case 2:
+		return unsafe.Pointer(unsafe.SliceData(make([]uint16, ceiling(size, align))))
+	case 4:
+		return unsafe.Pointer(unsafe.SliceData(make([]uint32, ceiling(size, align))))
+	case 8:
+		return unsafe.Pointer(unsafe.SliceData(make([]uint64, ceiling(size, align))))
+	default:
+		panic(fmt.Sprintf("unsupported alignment: %v", align))
+	}
+}
+
+// NB: `cabi_realloc` may be called before the Go runtime has been initialized,
+// in which case we need to use `runtime.sbrk` to do allocations.  The following
+// is an abbreviation of [Till's
+// efforts](https://github.com/bytecodealliance/go-modules/pull/367).
+
+//go:linkname sbrk runtime.sbrk
+func sbrk(n uintptr) unsafe.Pointer
+
+var useGCAllocations = false
+
+func init() {
+	useGCAllocations = true
+}
+
+func offset(ptr, align uintptr) uintptr {
+	newptr := (ptr + align - 1) &^ (align - 1)
```

Comment:
Is `&^` a typo? If so I'm curious, what does it do?

Reply:
I stole that code from @tschneidereit's [PR](https://github.com/bytecodealliance/go-modules/pull/367) :shrug:

Reply:
`&^` is `AND NOT`, i.e. this aligns the pointer, and the function returns offset between the aligned and unaligned addresses. As the name "offset" clearly says, obviously. (Sorry for the poor naming and lack of documentation!)